Migrate ReportActionCompose to function

stas-kravchuk opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->
Refactoring the `ReportActionCompose.js` component from class to functional one.

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ https://github.com/Expensify/App/issues/16263


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

-  Write and send a message
- Insert emoji from emoji picker in empty input
- Type some message and try to insert emoji into different places where the cursor is placed (with different selections)
- Insert emoji with emoji suggestion (with `:` )
- Type huge text to test composer size change with different amounts of lines
- Add/send an attachment
- Verify that the focus on the text input is being saved after the attachment modal is open, emoji picker is open, the message sending, etc
- Click on the arrows button to see the composer in expand/collapsed view (full size mode)
- Add mention, send message with mention

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

None needed.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

Same as in "Tests"

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

![web 1](https://github.com/Expensify/App/assets/40563919/0a971f8c-12f1-4c57-a0ea-f382de29ddb6)

![web 2](https://github.com/Expensify/App/assets/40563919/f8aae76d-ad0b-4da3-8220-58399118ba24)

![web 3](https://github.com/Expensify/App/assets/40563919/40ddacb1-a06f-4e5f-94f6-7ba9dbe29dfb)


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Chrome</summary>

![android web 1](https://github.com/Expensify/App/assets/40563919/3abf6058-bf5e-4d27-90f4-326f6cfe11d4)

![android web 2](https://github.com/Expensify/App/assets/40563919/77532d2c-8bfa-46c0-bc80-9a44abecdc40)

![android web 3](https://github.com/Expensify/App/assets/40563919/2156717b-56e4-4583-9a84-1470bbfab011)


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>

![ios web 1](https://github.com/Expensify/App/assets/40563919/c4f2ad18-c258-462e-bd49-81f2b961cd91)

![ios web 2](https://github.com/Expensify/App/assets/40563919/aa3a2430-b02c-4aca-9bd2-5d30f5bbf0a1)

![ios web 3](https://github.com/Expensify/App/assets/40563919/a8f6c796-4441-4000-b7d3-68525e031977)


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Desktop</summary>

![desktop 1](https://github.com/Expensify/App/assets/40563919/38a44f7b-5b2e-45b7-aca1-530adef8c8cc)

![desktop 2](https://github.com/Expensify/App/assets/40563919/b5ee7ba4-b7ac-4a74-8f95-7924e4a21ad8)

![desktop 3](https://github.com/Expensify/App/assets/40563919/afbf0cce-361e-4606-9d38-1d8a3f3d5808)


<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS</summary>

![ios 1](https://github.com/Expensify/App/assets/40563919/04951ba6-dc62-48da-8b0e-7ffdc7295a9f)

![ios 2](https://github.com/Expensify/App/assets/40563919/60f0fb3f-0ff3-4846-b520-2e734a3d124f)

![ios 3](https://github.com/Expensify/App/assets/40563919/1339f176-1238-4f07-b5bf-9b5ba8cded9b)


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Android</summary>

![android 1](https://github.com/Expensify/App/assets/40563919/16d1a249-4edd-4401-b2fa-294a9c1be77a)

![android 2](https://github.com/Expensify/App/assets/40563919/de21b2d6-a361-4810-8ce8-60683d5ecca6)

![android 3](https://github.com/Expensify/App/assets/40563919/f9c06afa-ddb8-4a58-a1b7-96d811f16c88)

<!-- add screenshots or videos here -->

</details>

stas-kravchuk committed:
refactor: in-progress - migrate ReportActionCompose from class to function component

stas-kravchuk committed:
refactor: in-progress - add few comments, minor changes in callbacks dependencies

stas-kravchuk committed:
refactor: WIP - move few functions out of component, replace some useCallback with use memo, fix bug with emoji show only after 3 symbols

stas-kravchuk committed:
refactor: WIP - move code from ReportActionComposeF to ReportAction compose (replace class file with function one)

stas-kravchuk committed:
refactor: WIP - remove prevProps useref - replace it with useffect logic, add few TODO comment, remove dev file ReportActionComposeF

stas-kravchuk committed:
refactor: rewrite few const and callback to useMemo

stas-kravchuk committed:
merge main into branch

stas-kravchuk committed:
refactor: prettier changes, moved onDrop event to separate useCallback

stas-kravchuk committed:
merge main into branch

stas-kravchuk committed:
merge main into branch

stas-kravchuk committed:
fix: return personalDetails onyx prop after merge conflict fix

stas-kravchuk committed:
fix: fix bug with ENTER press when suggestions exist, prettier style fix

stas-kravchuk committed:
merge main into branch

stas-kravchuk committed:
refactor: change isEmptyChat to useMemo from useCallback

stas-kravchuk committed:
merge main into branch

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -159,1008 +172,974 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
+            </OfflineWithFeedback>
+            {isDraggingOver && <ReportDropUI />}
+            {!_.isEmpty(suggestionValues.suggestedEmojis) && suggestionValues.shouldShowEmojiSuggestionMenu && (
+                <ArrowKeyFocusManager
```
We added a `useArrowKeyFocusManager` hook, let's use that instead. It's a bit simpler.

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
fix: import

Commit:
use "useKeyboardShortcut" instead of manually subscribing

Commit:
simplify arrow key handling

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
fix: 2 minor issues

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

stas-kravchuk commented:
I have read the CLA Document and I hereby sign the CLA

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented:
@roryabraham Resolved all merge conflicts, so it's ready to review + merge again!

Commit:
fix: remaining issues

Commit:
add remaining hook dependencies

Commit:
fix: prettier

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented:
Done! @0xmiroslav

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -155,1060 +168,1047 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
+    );
+
+    // If we are on a small width device then don't show last 3 items from conciergePlaceholderOptions
+    const conciergePlaceholderRandomIndex = useMemo(() => _.random(props.translate('reportActionCompose.conciergePlaceholderOptions').length - (props.isSmallScreenWidth ? 4 : 1)), [props]);
```
I suppose once `props.isSmallScreenWidth` or `reportActionCompose.conciergePlaceholderOptions` would change (by a `props.translate` change), we would update this value too?

chrispader commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -155,1060 +168,1047 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
-            },
-        ];
-    }
+    useKeyboardShortcut(
```
you're right. This seems to be a left-over from previous changes. Thanks!

Commit:
extract translate from props

Commit:
fix: minor issues

Commit:
remove useKeyboardShortcut import

Commit:
update conciergePlaceholderRandomIndex only on mount

chrispader commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -155,1060 +168,1047 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
+                    shouldIncludeReportRecipientLocalTimeHeight={shouldShowReportRecipientLocalTime}
+                />
+            )}
+            {!_.isEmpty(suggestionValues.suggestedMentions) && suggestionValues.shouldShowMentionSuggestionMenu && (
```
Thanks! Added the condition back

Commit:
fix: emoji suggestions

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
fix: useArrowKeyFocusManager not working

chrispader commented:
@0xmiroslav @roryabraham i think we should wait for the changes from https://github.com/Expensify/App/issues/19700#issuecomment-1597120028

> @chrispader please fix conflict by applying changes from #21030.
> I think that PR is the only one which touched this component after last pull from main.

There were a few more changes than from that one PR, just merged them!

Also the merge of [this PR](https://github.com/Expensify/App/pull/20756) helped me solve the problem with multiple `useArrowKeyFocusManager()` calls, as we can now only enable them if the emoji/mention suggestion menu is visible

chrispader commented:
Still investigating why mentions are not working in multiline inputs...

Commit:
make consistent with main

Commit:
fix: mentions not working in multiline

chrispader commented:
> Still investigating why mentions are not working in multiline inputs...

Fixed ✅ @0xmiroslav

roryabraham commented:
@0xmiroslav please prioritize reviewing this soon after they resolve conflicts, because this component keeps having lots of conflicts

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented:
@0xmiroslav conflicts resolved!

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented:
> conflicts again

resolved ✅

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
fix wrong prop

Commit:
update package-lock

Commit:
fix: arrow up to edit not working

chrispader commented:
@0xmiroslav arrow up issue is fixed. Looking into the performance regression right now

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

roryabraham commented:
We've got conflicts again, then bump @0xmiroslav for review

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented:
@0xmiroslav conflicts resolved and performance regression should also be fixed!

https://github.com/Expensify/App/assets/20173411/9d3a2732-4d81-4157-9563-5d5b1b2fc6b4

chrispader commented:
I noticed keyboard performance is also slower during recording and especially while the "macOS Keyboard Viewer" is opened.

So IRL, this should be more performant than in the recordings...

Here is a clip without the Keyboard Viewer open:

https://github.com/Expensify/App/assets/20173411/7cc56280-fe05-457d-8a45-680768543747

chrispader commented:
@roryabraham i'm not the issue creator, so unfortunately i can't remove the `[WIP]` from the title. This is ready review again 👍

chrispader committed:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader committed:
fix: dependency array

chrispader commented:
@0xmiroslav @roryabraham conflicts resolved!

chrispader commented:
> I performed test steps in recent PRs and confirm that they're not reverted.

What do you mean by that? Are you talking about the performance regression?

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented:
@0xmiroslav fixed conflicts! 👍

chrispader commented:
> Is it possible to add some performance measurement unit test while typing?

I can def give it a look. @roryabraham wdyt?

roryabraham commented:
Let's save the automated typing performance tests for a follow-up. I like the idea but this PR has been delayed / conflicted enough as-is

chrispader commented:
@0xmiroslav do you have some time to review this PR once more, before there are any new conflicts 😬 thanks!

chrispader committed:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented:
@0xmiroslav conflicts resolved 👍

chrispader committed:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented:
@0xmiroslav just resolved some more conflicts 😬

roryabraham commented:
@0xmiroslav bump for reviews

chrispader committed:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader committed:
fix: usePrevious calls

chrispader committed:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

ygshbht committed:
MacOS: Add 'Paste as Plain Text' to context menu

ygshbht committed:
MacOS: Update keyboard shortcuts for paste operations in context menu

ygshbht committed:
MacOS: Add accelerator for 'Paste and Match Style' in app menu

ygshbht committed:
Desktop: Add 'Paste as Plain Text' to macOS Edit menu with translations

ygshbht committed:
Desktop: Change hardcoded accelerator for pasteAsPlainText to a constant

chrispader commented:
@0xmiroslav @roryabraham conflicts again, going to fix this now!

@0xmiroslav could you review this by today/tomorrow?

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

ygshbht committed:
Lint code: Apply suggestions from code review

Co-authored-by: Manan <manan.jadhav@gmail.com>

ygshbht committed:
More code linting

ygshbht committed:
Desktop: Update accelerator for 'Paste and Match Style' to use constant

ygshbht committed:
Desktop: Refactor - Extract 'paste as plain text' logic to separate function

ygshbht committed:
Update 'Paste as Plain Text' menu item visibility and action

ygshbht committed:
Desktop: Add constant for 'Paste as Plain Text' menu item ID

ygshbht committed:
Desktop: Refactoring

ygshbht committed:
Desktop: Formatting

Krishna2323 committed:
fix: Wrong position of emoji picker modal

Signed-off-by: Krishna Gupta <belivethatkg@gmail.com>

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Commit:
fix: use prop destructuring

danieldoglas commented:
@0xmiroslav can you please prioritize reviewing this? Thanks

danieldoglas commented:
@0xmiroslav When you review this, we will need to solve conflicts again. So please test the branch fully, we'll resolve the conflicts and we do another test to confirm. This is the 3/4th time solving conflicts so try to make this the last one

Commit:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

chrispader commented:
> @chrispader can you please fix last conflict?

done! 👍

Commit:
fix: commentRef value not referenced

Commit:
remove console.log

Commit:
apply changes from #19011

chrispader commented:
> please apply changes from #19011. This is now non-modal popover

done! 👍

chrispader commented:
@roryabraham @0xmiroslav I'm gonna be on holiday until 20th of August. Just fixed the latest conflicts and implemented the #19011 PR @0xmiroslav requested.

If there's anything urgent, i'll hopefully be able to fix it on the go 👍

hannojg commented:
Also here to help with getting these changes merged in case more reviews come up

hannojg commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -165,1165 +178,1167 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
+        shouldExcludeTextAreaNodes: false,
+    });
+
+    const insertedEmojis = useRef([]);
```
I would rename this to `insertedEmojisRef` to make obvious with what data structure we are dealing here (because we misused it later on, check further comments)

hannojg replied:
I fixed that :v:

hannojg commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -165,1165 +178,1167 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
-        }
-    }
+    const debouncedUpdateFrequentlyUsedEmojis = useCallback(() => {
+        User.updateFrequentlyUsedEmojis(EmojiUtils.getFrequentlyUsedEmojis(insertedEmojis));
```
I think this needs to use `.current` as its a ref

hannojg replied:
I fixed that :v:

hannojg commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -165,1165 +178,1167 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
+
+            if (!_.isEmpty(emojis)) {
+                User.updateFrequentlyUsedEmojis(EmojiUtils.getFrequentlyUsedEmojis(emojis));
+                insertedEmojis.current = [...insertedEmojis, ...emojis];
```
I think this needs to use `.current` as its a ref (I am also getting warnings in my browser when typing otherwise)

danieldoglas commented:
removing myself from this since @roryabraham is back today!

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

hannojg committed:
fix issues after merge

hannojg committed:
fix ref usage

hannojg commented:
Merged and fixed issues with the component. Please review @0xmiroslav @roryabraham, thank you guys 😊

situchan commented:
I will continue review based on [discussion](https://expensify.slack.com/archives/C02NK2DQWUX/p1691641008846119?thread_ts=1691488353.531539&cid=C02NK2DQWUX).
@hannojg please fix conflict again

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

hannojg committed:
reapply https://github.com/Expensify/App/pull/24223

hannojg commented:
@situchan please check again now

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

hannojg committed:
fix send button styles

hannojg committed:
fix re-focus once modal closes

hannojg committed:
reset index to zero

mimic the behaviour of: https://github.com/Expensify/App/blob/dd0d98fa6868bba60e81bfdf66e5cdf7d2adaa37/src/pages/home/report/ReportActionCompose.js#L594-L607

hannojg commented:
Hey, thanks for the report. Please test again, everything has been addressed!

@situchan

hannojg committed:
remove debug statement

situchan commented:
https://github.com/Expensify/App/pull/23258 is just merged. Let's apply changes from https://github.com/Expensify/App/pull/23258/files#diff-7fa482b2895f4e36a2a3ad8233496489c6de1256a8c992524e2fc072aaba43f8

hannojg commented:
I found one bug where the draft doesn't sync between tabs. Working on a fix

hannojg commented:
Yeah, will add those changes you mentioned as well. In general it would be dope if we could hold all other PRs concerning this component until this one here is merged. We are trying to get it merged since three month, but always other PRs get handled first, thus creating conflicts in this PR :| 

(never mind me crying around, will apply the fix in a moment, thanks for testing this with me, lets get it merged 💪 )

hannojg committed:
fix comment darft tab synchronisation

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

hannojg committed:
reapply https://github.com/Expensify/App/pull/23258

situchan reviewed (changes requested):
> Bug3: [web] app crashes when mention someone picked from suggestions list (not 1st one) and press backspace

I am not able to find consistent reproducible step but similar crash still happens.

<img width="1792" alt="Screenshot 2023-08-14 at 7 44 48 AM" src="https://github.com/Expensify/App/assets/108292595/abc1c1e8-db21-40af-8c7a-a53e601db982">

I think this was what I was typing (with combination of arrow keys in suggestions list):

<img width="767" alt="Screenshot 2023-08-14 at 7 47 07 AM" src="https://github.com/Expensify/App/assets/108292595/d3162193-8b97-4716-bdb2-763352aefc85">

hannojg commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -166,1132 +179,1146 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
+                ...prevState,
+                ...nextState,
+            }));
+            setHighlightedMentionIndex(0);
```
You are right! This will also fix [the crash you were reporting earlier](https://github.com/Expensify/App/pull/18648#pullrequestreview-1575939289)!

hannojg committed:
Merge branch 'main' of github.com:Expensify/App into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

hannojg committed:
reapply https://github.com/Expensify/App/commit/1b9d86952111f755d5fe698546170c5fe6f10429

hannojg committed:
reset emoji index, fix crash

hannojg committed:
Update src/pages/home/report/ReportActionCompose.js

Co-authored-by: Situ Chandra Shil <108292595+situchan@users.noreply.github.com>

hannojg committed:
rename callback

hannojg committed:
remove deprecated comment

hannojg committed:
remove redundant code

hannojg committed:
clearer naming

hannojg committed:
remove debounce from showPopoverMenu

hannojg commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -166,1132 +179,1146 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
-        }
+    const showPopoverMenu = useMemo(
+        () =>
+            _.debounce(() => {
```
you are right, it isn't needed!

hannojg commented:
@situchan I addressed all review points. Can you please check again?

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -21,7 +21,7 @@ import compose from '../../../libs/compose';
 import PopoverMenu from '../../../components/PopoverMenu';
 import withWindowDimensions, {windowDimensionsPropTypes} from '../../../components/withWindowDimensions';
```
We should use the hook equivalents for `withWindowDimensions` and `withLocalize` and remove the associated props.

roryabraham replied:
Same for:

- `withNetwork`
- `withKeyboardState`
- `withAnimatedRef`
- `Permissions` and the `betas` Onyx key
- `withNavigation`
- `withNavigationFocus`

roryabraham reviewed (commented):
Not going to block on those changes though. Let's get this merged then do a follow-up!

mountiny commented:
@situchan Are you able to finish the checklist on this one today?

situchan commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -166,1137 +179,1146 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
+                const lastReportAction = _.find([...reportActions, parentReportAction], (action) => ReportUtils.canEditReportAction(action));
+
+                if (lastReportAction !== -1 && lastReportAction) {
+                    Report.saveReportActionDraft(reportID, lastReportAction.reportActionID, _.last(lastReportAction.message).html);
```
Sorry I had to hold merge https://github.com/Expensify/App/pull/24544 for this PR.
Now the branch conflicts.
This change should be enough while fixing conflict.
@mountiny can anyone who has write permission to this branch help as @hannojg is OOO?

```suggestion
                    Report.saveReportActionDraft(reportID, lastReportAction, _.last(lastReportAction.message).html);
```

mountiny replied:
asked

Szymon20000 committed:
Update src/pages/home/report/ReportActionCompose.js

Co-authored-by: Situ Chandra Shil <108292595+situchan@users.noreply.github.com>

Szymon20000 committed:
Merge branch 'main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Szymon20000 committed:
Update ReportActionCompose.js

Szymon20000 committed:
Update ReportActionCompose.js

Commit:
prettier signed

situchan commented:
Crash in android: (confirmed not happening on main)



https://github.com/Expensify/App/assets/108292595/2d286eb8-a5f9-4984-8a55-19d5f4469976

situchan commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -166,1137 +179,1146 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
+            // will be added after the user starts typing again on the keyboard. This package is
+            // a workaround to reset the keyboard natively.
+            if (RNTextInputReset) {
+                RNTextInputReset.resetKeyboardInput(findNodeHandle(textInputRef));
```
found root cause of crash. good easy-to-miss case
```suggestion
                RNTextInputReset.resetKeyboardInput(findNodeHandle(textInputRef.current));
```

situchan replied:
bump on applying this. And then all good!

Szymon20000 committed:
Update src/pages/home/report/ReportActionCompose.js

Co-authored-by: Situ Chandra Shil <108292595+situchan@users.noreply.github.com>

situchan reviewed (approved):
Checklist: https://github.com/Expensify/App/pull/18648#issuecomment-1680905876
We already did tons of tests here but there might still be edge cases we missed.
As this is the most important part in the app, I suggest QA team to do full regression test related to composer.

All yours @roryabraham @mountiny

roryabraham commented:
Can't merge due to unsigned commits

situchan commented:
Seems last 2 commits should be reverted and force push with signed commit

<img width="1253" alt="Screenshot 2023-08-18 at 12 29 34 AM" src="https://github.com/Expensify/App/assets/108292595/5c98b929-8b8a-4ad7-acb4-a5ecbf053740">


cc: @Szymon20000

situchan commented:
There should be conflict since https://github.com/Expensify/App/pull/24512 was merged

<img width="1006" alt="Screenshot 2023-08-18 at 9 01 43 PM" src="https://github.com/Expensify/App/assets/108292595/91085d7c-65cd-421e-8c5f-be6f4e8f579f">

mountiny commented:
Oh interesting its not showing the specifc files, I can see it now

Szymon20000 committed:
Merge remote-tracking branch 'origin/main' into @stas-kravchuk/16263-migrate-reportactioncompose-to-function

Szymon20000 committed:
linting

Szymon20000 committed:
linting

Szymon20000 committed:
linting 3

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5931749076) to staging by https://github.com/roryabraham in version: 1.3.56-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5958899174) to production by https://github.com/roryabraham in version: 1.3.56-24 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

roryabraham commented:
Not sure – @ginsuma can you please report the problem in #expensify-bugs?

roryabraham commented on src/pages/home/report/ReportActionCompose.js:
```diff
@@ -166,1133 +179,1140 @@ const getMaxArrowIndex = (numRows, isAutoSuggestionPickerLarge) => {
+
+    const onSelectionChange = useCallback(
+        (e) => {
+            LayoutAnimation.configureNext(LayoutAnimation.create(50, LayoutAnimation.Types.easeInEaseOut, LayoutAnimation.Properties.opacity));
```
As soon as Reanimated 3.5.0 lands we should start using Reanimated for LayoutAnimations going forward
